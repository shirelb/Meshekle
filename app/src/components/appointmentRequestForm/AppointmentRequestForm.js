import React, {Component} from 'react';
import {Modal, StyleSheet, Text, TouchableHighlight, View} from "react-native";
import {CheckBox, FormInput, FormLabel, FormValidationMessage} from "react-native-elements";
import {    SelectMultipleGroupButton} from "react-native-selectmultiple-button";

export default class AppointmentRequestForm extends Component {
    constructor(props) {
        super(props);

        this.subjects = [
            { value: "פן" },
            { value: "צבע" },
            { value: "תספורת" },
            { value: "החלקה" },
            { value: "גוונים" }
        ];

        this.state = {
            modalVisible: this.props.modalVisible,
            serviceProvider: this.props.serviceProvider,
            daysSelected: [],
            subjectSelected:[],
            subjectText: "",
            displaySubjectList: false,
            notes:'',
        };
    }

    componentWillReceiveProps(nextProps, nextContext) {
        this.setState({
            modalVisible: this.props.modalVisible,
            serviceProvider: this.props.serviceProvider,
        })
    }

    setModalVisible(visible) {
        this.setState({modalVisible: visible});
    }

    groupButtonOnSelectedValuesChange(selectedValues) {
        this.setState({
            subjectSelected: selectedValues
        });
    }

    render() {
        return (
            <View style={{marginTop: 22}}>
                <Modal
                    animationType="slide"
                    transparent={false}
                    visible={this.state.modalVisible}
                    onRequestClose={() => {
                        Alert.alert('Modal has been closed.');
                    }}>
                    <View style={{marginTop: 22}}>
                        <View>
                            <Text>Hello World!</Text>

                            {/*<FormLabel>Name</FormLabel>*/}
                            {/*<FormInput onChangeText={someFunction}/>*/}
                            {/*<FormValidationMessage>Error message</FormValidationMessage>*/}

                            <CheckBox
                                right
                                title='ראשון'
                                iconRight
                                iconType='material'
                                checkedIcon='clear'
                                uncheckedIcon='add'
                                checkedColor='red'
                                checked={this.state.checked}
                            />

                            <Text style={{ color: '#007AFF', marginLeft: 10 }}>
                                נושא  {this.state.subjectSelected.join( ", ")}
                            </Text>
                            <SelectMultipleGroupButton
                                containerViewStyle={{
                                    justifyContent: "flex-start"
                                }}
                                highLightStyle={{
                                    borderColor: "gray",
                                    backgroundColor: "transparent",
                                    textColor: "gray",
                                    borderTintColor: '#007AFF',
                                    backgroundTintColor: "transparent",
                                    textTintColor: '#007AFF',
                                }}
                                onSelectedValuesChange={selectedValues =>
                                    this.groupButtonOnSelectedValuesChange(selectedValues)
                                }
                                group={this.subjects}
                            />

                            <TouchableHighlight
                                onPress={() => {
                                    this.setModalVisible(!this.state.modalVisible);
                                }}>
                                <Text>Hide Modal</Text>
                            </TouchableHighlight>
                        </View>
                    </View>
                </Modal>

            </View>
        );
    }
}

const styles = StyleSheet.create({
});
