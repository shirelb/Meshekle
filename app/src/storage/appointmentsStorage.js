import axios from "axios";
import {APP_SOCKET, SERVER_URL} from "../shared/constants";


var getUserAppointmentRequests = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointmentRequests/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'requested'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointment requests error ', error)
        });
};

var getUserAppointments = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {
                status: 'set'
            },
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments error ', error)
        });
};

var getUserAppointmentById = function (userId, userHeaders, eventId) {
    return axios.get(`${SERVER_URL}/api/appointments/user/userId/${userId}`,
        {
            headers: userHeaders,
            params: {status: 'set', appointmentId: eventId}
        })
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user appointments by id error ', error)
        });
};

var postUserAppointmentRequest = function (userId, serviceProvider, appointmentRequest, userHeaders) {
    return axios.post(`${SERVER_URL}/api/appointmentRequests/user/request`,
        {
            userId: userId,
            serviceProviderId: serviceProvider.serviceProviderId,
            role: serviceProvider.role,
            availableTime: appointmentRequest.availableTime,
            notes: appointmentRequest.notes ? appointmentRequest.notes : '',
            subject: JSON.stringify(appointmentRequest.subject)
        },
        {
            headers: userHeaders,
        })
        .then(response => {
            APP_SOCKET.emit('userPostAppointmentRequests', {
                serviceProviderId: serviceProvider.serviceProviderId
            });

            return response;
        })
        .catch(error => {
            console.log('post user appointment request error ', error)
        });
};

var cancelAppointmentRequestById = (appointmentRequest, headers) => {
    return axios.put(`${SERVER_URL}/api/appointmentRequests/user/reject`,
        {
            userId: appointmentRequest.AppointmentDetail.clientId,
            appointmentRequestId: appointmentRequest.requestId,
        },
        {
            headers: headers,
            params: {
                status: 'rejected'
            },
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointmentRequests', {
                serviceProviderId: appointmentRequest.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('reject appointment request error ', error);
        });
};

var cancelAppointmentById = (appointment, headers) => {
    return axios.put(`${SERVER_URL}/api/appointments/user/cancel/userId/${appointment.AppointmentDetail.clientId}/appointmentId/${appointment.appointmentId}`,
        {},
        {
            headers: headers
        }
    )
        .then((response) => {
            APP_SOCKET.emit('userCancelAppointment', {
                serviceProviderId: appointment.AppointmentDetail.serviceProviderId,
            });

            return response
        })
        .catch((error) => {
            console.log('cancel appointment error ', error);
        });
};


export default {
    getUserAppointments,
    getUserAppointmentById,
    postUserAppointmentRequest,
    getUserAppointmentRequests,
    cancelAppointmentRequestById,
    cancelAppointmentById,
};
