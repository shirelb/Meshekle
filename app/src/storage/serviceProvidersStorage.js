import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getServiceProviders = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/serviceProviders`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('load appointments error ', error)
        });
};


export default {getServiceProviders};
