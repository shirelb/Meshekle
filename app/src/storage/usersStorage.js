import axios from "axios";
import {SERVER_URL} from "../shared/constants";

var getUserEvents = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/events/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user events error ', error)
        });
};

var getUserById = function (userId, userHeaders) {
    return axios.get(`${SERVER_URL}/api/users/userId/${userId}`,
        {headers: userHeaders}
    )
        .then(response => {
            return response;
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var getUsers = function (userHeaders) {
    return axios.get(`${SERVER_URL}/api/users`,
        {headers: userHeaders}
    )
        .then(response => {
            return response.data.filter(user => user.userId !== 1 && user.userId !== "1");
        })
        .catch(error => {
            console.log('get user by id error ', error)
        });
};

var userValidToken = function (token) {
    return axios.post(`${SERVER_URL}/api/users/validToken`,
        {
            "token": token,
        },
    );
}

var userLogin = function (userId, password) {
    return axios.post(`${SERVER_URL}/api/users/login/authenticate`,
        {
            "userId": userId,
            "password": password
        },
    );
}


export default {getUserEvents, getUserById, userValidToken, userLogin, getUsers};
