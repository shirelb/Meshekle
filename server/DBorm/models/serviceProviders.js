module.exports = (sequelize, type) => {
    return sequelize.define('ServiceProviders', {
        serviceProviderId:{
            type: type.INTEGER,
            primaryKey: true,
            autoIncrement: true,
            allowNull: false,
            required: true
        },
        role : {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        userId: {
            type: type.INTEGER,
            allowNull: false,
            required: true
        },
        operationTime: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
        phoneNumber: {
            type: type.STRING,
            allowNull: false,
            required: true
        },
    });
};