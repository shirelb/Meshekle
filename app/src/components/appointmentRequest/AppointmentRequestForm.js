import React, {Component} from 'react';
import {Alert, Modal, ScrollView, StyleSheet, View} from "react-native";
import {CheckBox, FormInput, FormLabel, FormValidationMessage, Icon, Text} from "react-native-elements";
import {SelectMultipleGroupButton} from "react-native-selectmultiple-button";
import DateTimePicker from 'react-native-modal-datetime-picker';
import Button from "../submitButton/Button";
import {List} from "react-native-paper";
import moment from 'moment';
import appointmentsStorage from "../../storage/appointmentsStorage";


export default class AppointmentRequestForm extends Component {
    constructor(props) {
        super(props);

        this.subjects = [];

        this.state = {
            modalVisible: this.props.modalVisible,
            datesAndHoursSelected: this.props.selectedDate === '' || typeof this.props.selectedDate !== "string" ? [] : [{
                'date': this.props.selectedDate,
                'hours': [{startHour: "", endHour: ""}, {startHour: "", endHour: ""}],
                'expanded': false,
            }],
            subjectSelected: [],
            subjectText: "",
            displaySubjectList: false,
            notes: '',
            isDateTimePickerVisible: false,
            isStartDateTimePickerVisible: false,
            isEndDateTimePickerVisible: false,
            dateClicked: new Date(),
            startTimeClicked: '',
            endTimeClicked: '',
            errorMsg: '',
            errorVisible: true
        };

        this.handleDatePicked = this.handleDatePicked.bind(this);
        this.handleStartTimePicked = this.handleStartTimePicked.bind(this);
        this.handleEndTimePicked = this.handleEndTimePicked.bind(this);
    }

    componentWillReceiveProps(nextProps, nextContext) {
        this.setState({
            modalVisible: nextProps.modalVisible,
            serviceProvider: nextProps.serviceProvider,

            datesAndHoursSelected: nextProps.selectedDate === '' || typeof nextProps.selectedDate !== "string" ? [] : [{
                'date': nextProps.selectedDate,
                'hours': [{startHour: "", endHour: ""}, {startHour: "", endHour: ""}],
                'expanded': false,
            }],
            subjectSelected: [],
            subjectText: "",
            displaySubjectList: false,
            notes: '',
            isDateTimePickerVisible: false,
            isStartDateTimePickerVisible: false,
            isEndDateTimePickerVisible: false,
            dateClicked: new Date(),
            startTimeClicked: '',
            endTimeClicked: '',
            errorMsg: '',
            errorVisible: true
        });

        this.subjects = [];
        nextProps.serviceProvider.subjects ?
            JSON.parse(nextProps.serviceProvider.subjects).map(subject => {
                this.subjects.push({value: subject})
            })
            : null;
    }

    setModalVisible(visible) {
        this.setState({modalVisible: visible});
        this.props.closeAppointmentRequestForm();
    }

    groupButtonOnSelectedValuesChange(selectedValues) {
        this.setState({
            subjectSelected: selectedValues,
            errorVisible: false
        });
    }

    handleDatePicked = (selectedDate) => {
        const date = moment(selectedDate).format('YYYY-MM-DD');
        let datestimes = this.state.datesAndHoursSelected;
        let found = datestimes.some(function (el) {
            return el.date === date;
        });
        if (!found) {
            datestimes.push({'date': date, 'hours': [], 'expanded': false});
            // let expanded = this.state.expanded;
            // expanded.push({date: false});
            this.setState({
                datesAndHoursSelected: datestimes,
                // expanded: expanded,
            });
        }
        this.setState({
            isDateTimePickerVisible: false,
            isStartDateTimePickerVisible: true,
            dateClicked: date,
        });
    };

    handleStartTimePicked = (selectedTime) => {
        const time = moment(selectedTime).format('HH:mm');
        let datestimes = this.state.datesAndHoursSelected;
        datestimes.forEach(dateTime => {
            if (dateTime.date === this.state.dateClicked)
                if (!dateTime.hours)
                    dateTime.hours = [];
            dateTime.hours.push({'startHour': time});
        });

        this.setState({
            isStartDateTimePickerVisible: false,
            isEndDateTimePickerVisible: true,
            datesAndHoursSelected: datestimes,
            startTimeClicked: time,
        });
    };

    handleEndTimePicked = (selectedTime) => {
        const time = moment(selectedTime).format('HH:mm');
        let datestimes = this.state.datesAndHoursSelected;
        datestimes.forEach(dateTime => {
            if (dateTime.date === this.state.dateClicked)
                dateTime['hours'].forEach(times => {
                    if (times.startHour === this.state.startTimeClicked)
                        times['endHour'] = time;
                })
        });

        datestimes.forEach(dateTime => {
            dateTime['hours'] = dateTime['hours'].filter(function (times) {
                return times.startHour && times.endHour;
            });
        });

        this.setState({
            isEndDateTimePickerVisible: false,
            datesAndHoursSelected: datestimes,
            endTimeClicked: time,
        });
    };

    sendAppointmentRequest() {
        if (this.state.datesAndHoursSelected.length === 0 ||
            this.state.subjectSelected.length === 0) {
            this.setState({errorMsg: 'ישנו מידע חסר, השלם שדות חובה (שדות עם *)', errorVisible: true})
        } else {
            this.setModalVisible(!this.state.modalVisible);

            let appointmentRequest = {
                availableTime: this.state.datesAndHoursSelected,
                notes: this.state.notes,
                subject: this.state.subjectSelected,
            };

            appointmentsStorage.postUserAppointmentRequest(this.props.userId, this.props.serviceProvider, appointmentRequest, this.props.userHeaders)
                .then(() => {
                    Alert.alert(
                        'התראה',
                        'הבקשה נשלחה בהצלחה',
                    );
                })
        }
    }

    deleteHoursSelected = (item, hourIndex, dateIndex) => {
        let updateDatesAndHoursSelected = this.state.datesAndHoursSelected;
        updateDatesAndHoursSelected[dateIndex].hours.splice(hourIndex, 1);

        if (updateDatesAndHoursSelected[dateIndex].hours.length === 0)
            updateDatesAndHoursSelected.splice(dateIndex, 1);

        this.setState({
            datesAndHoursSelected: updateDatesAndHoursSelected,
        })
    };


    render() {
        return (
            <Modal
                animationType="slide"
                transparent={false}
                visible={this.state.modalVisible}
                onRequestClose={() => {
                    this.setModalVisible(!this.state.modalVisible);
                }}
            >
                <View style={{marginTop: 20}}>
                    <ScrollView>
                        <Text h4 style={styles.textTitle}>בקשת תור</Text>
                        {/*<Header
                                backgroundColor={'white'}
                                // leftComponent={{ icon: 'menu', color: '#fff' }}
                                centerComponent={{ text: 'בקשת תור', style: { color: '#050505',fontWeight: 'bold',fontSize:26 } }}
                                // rightComponent={{ icon: 'home', color: '#fff' }}
                            />*/}

                        <FormLabel>*נותן שירות</FormLabel>
                        <Text style={{marginLeft: 10}}>
                            {this.props.serviceProvider.fullname}
                        </Text>

                        <FormLabel> *ענף</FormLabel>
                        <Text style={{marginLeft: 10}}>
                            {this.props.serviceProvider.role}
                        </Text>

                        <FormLabel> *תאריכים ושעות אופציונאליים</FormLabel>
                        <List.Section title={this.state.selectedDate}>
                            {Array.isArray(this.state.datesAndHoursSelected) &&
                            this.state.datesAndHoursSelected.map((item, index) => {
                                return <List.Accordion
                                    key={index}
                                    // title={moment(item.startDateAndTime).format('HH:mm') + '-' + moment(item.endDateAndTime).format('HH:mm')}
                                    title={item.date}
                                    // description={item.AppointmentDetail.role + ',' + item.AppointmentDetail.serviceProviderId}
                                    // left={props => <List.Icon {...props} icon="perm-contact-calendar"/>}
                                    // expanded={this.state.expanded[item.date]}
                                    expanded={item.expanded}
                                    onPress={() => {
                                        // let expanded = this.state.expanded;
                                        // expanded[item.date] = !expanded[item.date];
                                        // this.setState({expanded: expanded})
                                        let datesAndTimes = this.state.datesAndHoursSelected;
                                        datesAndTimes[index].expanded = !datesAndTimes[index].expanded;
                                        this.setState({datesAndHoursSelected: datesAndTimes})
                                    }}
                                >
                                    {Array.isArray(item['hours']) && item['hours'] ?
                                        item['hours'].map((hour, j) => {
                                            return <List.Item
                                                key={j}
                                                title={hour.startHour + '-' + hour.endHour}
                                                // description={item.remarks}
                                                containerStyle={{borderBottomWidth: 0}}
                                                right={props => <Icon {...props}
                                                                      name="delete-forever"
                                                                      color={'red'}
                                                                      onPress={() => this.deleteHoursSelected(item, j, index)}/>}
                                            />
                                        }) : null
                                    }
                                </List.Accordion>
                            })
                            }
                        </List.Section>

                        <CheckBox
                            left
                            title='הוסף תאריך ושעות'
                            iconLeft
                            iconType='material'
                            checkedIcon='clear'
                            uncheckedIcon='add'
                            // checkedColor='red'
                            // checked={this.state.checked}
                            onPress={() => this.setState({
                                errorVisible: false,
                                isDateTimePickerVisible: true,
                                isStartDateTimePickerVisible: false,
                                isEndDateTimePickerVisible: false,
                            })}
                        />

                        {/*<TouchableOpacity onPress={() => this.setState({isDateTimePickerVisible: true})}>*/}
                        {/*<Text>בחר תאריך</Text>*/}
                        {/*</TouchableOpacity>*/}
                        <DateTimePicker
                            isVisible={this.state.isDateTimePickerVisible}
                            onConfirm={this.handleDatePicked}
                            onCancel={() => this.setState({isDateTimePickerVisible: false})}
                            is24Hour={true}
                            mode={'date'}
                            // datePickerModeAndroid={'spinner'}
                            title='תאריך'
                        />
                        {/*<TouchableOpacity onPress={() => this.setState({isStartDateTimePickerVisible: true})}>
                                <Text>בחר שעת התחלה</Text>
                            </TouchableOpacity>*/}
                        <DateTimePicker
                            date={new Date(this.state.dateClicked)}
                            isVisible={this.state.isStartDateTimePickerVisible}
                            onConfirm={this.handleStartTimePicked}
                            onCancel={() => this.setState({isStartDateTimePickerVisible: false})}
                            is24Hour={true}
                            mode={'time'}
                            title='זמן התחלה'
                        />
                        {/* <TouchableOpacity onPress={() => this.setState({isEndDateTimePickerVisible: true})}>
                                <Text>בחר שעת סיום</Text>
                            </TouchableOpacity>*/}
                        <DateTimePicker
                            date={new Date(this.state.dateClicked)}
                            isVisible={this.state.isEndDateTimePickerVisible}
                            onConfirm={this.handleEndTimePicked}
                            onCancel={() => this.setState({isEndDateTimePickerVisible: false})}
                            is24Hour={true}
                            mode={'time'}
                            title='זמן סיום'
                        />

                        <FormLabel>*נושא</FormLabel>
                        <Text style={{marginLeft: 10}}>
                            {this.state.subjectSelected.join(", ")}
                        </Text>
                        <SelectMultipleGroupButton
                            containerViewStyle={{
                                justifyContent: "flex-start"
                            }}
                            highLightStyle={{
                                borderColor: "gray",
                                backgroundColor: "transparent",
                                textColor: "gray",
                                borderTintColor: '#007AFF',
                                backgroundTintColor: "transparent",
                                textTintColor: '#007AFF',
                            }}
                            onSelectedValuesChange={selectedValues =>
                                this.groupButtonOnSelectedValuesChange(selectedValues)
                            }
                            group={this.subjects}
                        />

                        <FormLabel>הערות</FormLabel>
                        <FormInput
                            editable={true}
                            maxLength={40}
                            value={this.state.notes}
                            placeholder={"כתוב הערות נוספות כאן"}
                            multiline={true}
                            onChangeText={(text) => this.setState({notes: text})}
                            onFocus={() => this.setState({errorVisible: false})}
                        />

                        {
                            this.state.errorVisible === true ?
                                <FormValidationMessage>{this.state.errorMsg}</FormValidationMessage>
                                : null
                        }

                        <View style={{marginTop: 20}}>
                            <Button
                                label='שלח'
                                onPress={() => {
                                    this.sendAppointmentRequest();
                                }}
                            />

                            <Button
                                label='חזור'
                                onPress={() => {
                                    this.setModalVisible(!this.state.modalVisible);
                                }}
                            />
                        </View>

                    </ScrollView>
                </View>
            </Modal>
        );
    }
}

const styles = StyleSheet.create({
    textTitle: {
        textAlign: 'center',
        color: '#050505',
        fontWeight: 'bold',
    }
});
