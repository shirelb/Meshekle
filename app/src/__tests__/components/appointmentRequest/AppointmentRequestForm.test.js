import React from 'react';

import {Modal} from "react-native";
import {CheckBox, FormInput, FormLabel, FormValidationMessage, Text} from "react-native-elements";
import {List} from "react-native-paper";
import {shallow} from "enzyme/build";
import AppointmentRequestForm from "../../../components/appointmentRequest/AppointmentRequestForm";
import usersStorage from "../../../storage/usersStorage";
import phoneStorage from "react-native-simple-store";
import users from "../../jsons/users";
import serviceProvidersStorage from "../../../storage/serviceProvidersStorage";
import serviceProviders from "../../jsons/serviceProviders";
import appointmentsStorage from "../../../storage/appointmentsStorage";
import user013637605AppointmentRequests from "../../jsons/user013637605AppointmentRequests";

jest.mock("../../../storage/serviceProvidersStorage");
jest.mock("../../../storage/usersStorage");
jest.mock("react-native-simple-store");


describe("AppointmentRequestForm should", () => {
    let wrapper = null;
    let componentInstance = null;
    const serviceProviderTest = serviceProviders[0];
    const props = {
        modalVisible: true,
        serviceProvider: serviceProviderTest,
        closeAppointmentRequestForm: jest.fn(),
    };
    const mockStore = {
        userData: {
            serviceProviderId: "549963652",
            userId: "549963652",
            token: "some token"
        }
    };
    const navigation = {navigate: jest.fn(), dispatch: jest.fn()};

    phoneStorage.get = jest.fn().mockImplementation((key) => Promise.resolve(mockStore[key]));
    phoneStorage.update = jest.fn().mockImplementation((key, value) => Promise.resolve(mockStore[key] = value));
    usersStorage.getUsers = jest.fn().mockImplementation(() => Promise.resolve(users));
    usersStorage.getUserByUserID = jest.fn().mockImplementation((userId) => Promise.resolve(users.filter(user => user.userId === userId)[0]));
    serviceProvidersStorage.getServiceProviders = jest.fn().mockResolvedValue(serviceProviders);
    serviceProvidersStorage.getServiceProviderUserDetails = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve({data: users.filter(user => user.userId === serviceProviderId)[0]}));
    serviceProvidersStorage.getServiceProviderById = jest.fn().mockImplementation((serviceProviderId) => Promise.resolve(serviceProviders.filter(provider =>
        provider.serviceProviderId === serviceProviderId)));
    appointmentsStorage.getUserAppointmentRequests = jest.fn().mockImplementation(() => Promise.resolve(user013637605AppointmentRequests));
    appointmentsStorage.postUserAppointmentRequest = jest.fn().mockImplementation(() => Promise.resolve());

    beforeEach(() => {
        jest.clearAllMocks();
    });

    // xit('match snapshot', async () => {
    //     wrapper = shallow(<AppointmentRequestForm {...props} />);
    //     expect(wrapper).toMatchSnapshot();
    // });

    it("render what the user see", async () => {
        wrapper = await shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        await wrapper.update();

        expect(wrapper.find(Modal)).toHaveLength(1);
        expect(wrapper.find(Text)).toHaveLength(7);
        expect(wrapper.find(Text).at(0).props().children).toEqual('בקשת תור');
        expect(wrapper.find(Text).at(1).props().children).toEqual(serviceProviderTest.fullname);
        expect(wrapper.find(Text).at(2).props().children).toEqual(serviceProviderTest.role);
        expect(wrapper.find('FormLabel')).toHaveLength(6);
        expect(wrapper.find('FormLabel').at(0).props().children).toEqual('*נותן שירות');
        expect(wrapper.find('FormLabel').at(1).props().children).toEqual(" *ענף");
        expect(wrapper.find('FormLabel').at(2).props().children).toEqual(' *שעות פעילות');
        expect(wrapper.find('FormLabel').at(3).props().children).toEqual(' *תאריכים ושעות אופציונאליים');
        expect(wrapper.find('FormLabel').at(4).props().children).toEqual("*נושא");
        expect(wrapper.find('FormLabel').at(5).props().children).toEqual('הערות');
        expect(wrapper.find(List.Section)).toHaveLength(1);
        expect(wrapper.find(List.Accordion)).toHaveLength(0);
        expect(wrapper.find(List.Item)).toHaveLength(0);
        expect(wrapper.find('CheckBox')).toHaveLength(1);
        expect(wrapper.find('CheckBox').props().title).toEqual('הוסף תאריך ושעות');
        expect(wrapper.find('SelectMultipleGroupButton')).toHaveLength(1);
        expect(wrapper.find('FormInput')).toHaveLength(1);
        expect(wrapper.find('FormValidationMessage')).toHaveLength(0);
        expect(wrapper.find('Button')).toHaveLength(2);
        expect(wrapper.find('Button').at(0).props().label).toEqual('שלח');
        expect(wrapper.find('Button').at(1).props().label).toEqual('חזור');
    });

    it("mounted with the right data", async () => {
        expect(componentInstance.state.modalVisible).toEqual(true);
        expect(componentInstance.state.subjectSelected).toEqual([]);
        expect(componentInstance.state.subjectText).toEqual('');
        expect(componentInstance.state.displaySubjectList).toEqual(false);
        expect(componentInstance.state.notes).toEqual('');
        expect(componentInstance.state.isDateTimePickerVisible).toEqual(false);
        expect(componentInstance.state.isStartDateTimePickerVisible).toEqual(false);
        expect(componentInstance.state.isEndDateTimePickerVisible).toEqual(false);
        expect(componentInstance.state.startTimeClicked).toEqual('');
        expect(componentInstance.state.endTimeClicked).toEqual('');
        expect(componentInstance.state.errorMsg).toEqual('');
        expect(componentInstance.state.errorVisible).toEqual(false);
    });

    it("handle setModalVisible", async () => {
        wrapper = shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        expect(componentInstance.state.modalVisible).toEqual(true);

        await componentInstance.setModalVisible(false);
        expect(componentInstance.state.modalVisible).toEqual(false);
    });

    it("handle groupButtonOnSelectedValuesChange", async () => {
        wrapper = shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        await componentInstance.groupButtonOnSelectedValuesChange(["פן"]);
        expect(componentInstance.state.subjectSelected).toEqual(["פן"]);
        expect(componentInstance.state.errorVisible).toEqual(false);
    });

    it("delete Hours Selected", async () => {
        wrapper = shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        await componentInstance.setState({
            datesAndHoursSelected: [{
                'date': "2019-04-03",
                'hours': [{startHour: "14:44", endHour: "18:33"}, {startHour: "12:33", endHour: "14:44"}],
                'expanded': false,
            }],
        });

        await componentInstance.deleteHoursSelected({},0,0);
        expect(componentInstance.state.datesAndHoursSelected.length).toEqual(1);
        expect(componentInstance.state.datesAndHoursSelected[0].hours.length).toEqual(1);
    });

    it("not send Appointment Request with empty subjects", async () => {
        wrapper = shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        await componentInstance.setState({
            datesAndHoursSelected: [{
                'date': "2019-04-03",
                'hours': [{startHour: "14:44", endHour: "18:33"}, {startHour: "12:33", endHour: "14:44"}],
                'expanded': false,
            }],
            subjectSelected:[]
        });

        await wrapper.find('Button').at(0).props().onPress();
        expect(componentInstance.state.errorMsg).toEqual('ישנו מידע חסר, השלם שדות חובה (שדות עם *)');
        expect(componentInstance.state.errorVisible).toEqual(true);
    });

    it("not send Appointment Request", async () => {
        wrapper = shallow(<AppointmentRequestForm {...props} />);
        componentInstance = wrapper.instance();

        await componentInstance.setState({
            datesAndHoursSelected: [{
                'date': "2019-04-03",
                'hours': [{startHour: "14:44", endHour: "18:33"}, {startHour: "12:33", endHour: "14:44"}],
                'expanded': false,
            }],
            subjectSelected:["פן"]
        });

        await wrapper.find('Button').at(0).props().onPress();
        expect(appointmentsStorage.postUserAppointmentRequest).toHaveBeenCalledTimes(1);
    });

});
