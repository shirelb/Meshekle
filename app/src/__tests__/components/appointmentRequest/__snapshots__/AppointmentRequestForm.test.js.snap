// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentRequestForm should match snapshot 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={false}
  visible={true}
>
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <ScrollViewMock>
      <TextElement
        h4={true}
        style={
          Object {
            "color": "#050505",
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        בקשת תור
      </TextElement>
      <FormLabel>
        *נותן שירות
      </FormLabel>
      <TextElement
        style={
          Object {
            "marginLeft": 10,
          }
        }
      />
      <FormLabel>
         *ענף
      </FormLabel>
      <TextElement
        style={
          Object {
            "marginLeft": 10,
          }
        }
      >
        appointmentsHairDresser
      </TextElement>
      <FormLabel>
         *תאריכים ושעות אופציונאליים
      </FormLabel>
      <withTheme(List.Section) />
      <CheckBox
        center={false}
        checked={false}
        checkedColor="green"
        checkedIcon="clear"
        iconLeft={true}
        iconRight={false}
        iconType="material"
        left={true}
        onPress={[Function]}
        right={false}
        size={24}
        title="הוסף תאריך ושעות"
        uncheckedColor="#bfbfbf"
        uncheckedIcon="add"
      />
      <CustomDatePickerIOS
        cancelTextIOS="Cancel"
        confirmTextIOS="Confirm"
        date={2019-06-03T15:05:19.737Z}
        is24Hour={true}
        isVisible={false}
        mode="date"
        neverDisableConfirmIOS={false}
        onCancel={[Function]}
        onConfirm={[Function]}
        onDateChange={[Function]}
        onHideAfterConfirm={[Function]}
        reactNativeModalPropsIOS={Object {}}
        title="תאריך"
        titleIOS="Pick a date"
      />
      <CustomDatePickerIOS
        cancelTextIOS="Cancel"
        confirmTextIOS="Confirm"
        date={2019-06-03T15:06:39.303Z}
        is24Hour={true}
        isVisible={false}
        mode="time"
        neverDisableConfirmIOS={false}
        onCancel={[Function]}
        onConfirm={[Function]}
        onDateChange={[Function]}
        onHideAfterConfirm={[Function]}
        reactNativeModalPropsIOS={Object {}}
        title="זמן התחלה"
        titleIOS="Pick a date"
      />
      <CustomDatePickerIOS
        cancelTextIOS="Cancel"
        confirmTextIOS="Confirm"
        date={2019-06-03T15:06:39.303Z}
        is24Hour={true}
        isVisible={false}
        mode="time"
        neverDisableConfirmIOS={false}
        onCancel={[Function]}
        onConfirm={[Function]}
        onDateChange={[Function]}
        onHideAfterConfirm={[Function]}
        reactNativeModalPropsIOS={Object {}}
        title="זמן סיום"
        titleIOS="Pick a date"
      />
      <FormLabel>
        *נושא
      </FormLabel>
      <TextElement
        style={
          Object {
            "marginLeft": 10,
          }
        }
      />
      <SelectMultipleGroupButton
        containerViewStyle={
          Object {
            "justifyContent": "flex-start",
          }
        }
        group={Array []}
        highLightStyle={
          Object {
            "backgroundColor": "transparent",
            "backgroundTintColor": "transparent",
            "borderColor": "gray",
            "borderTintColor": "#007AFF",
            "textColor": "gray",
            "textTintColor": "#007AFF",
          }
        }
        multiple={true}
        onSelectedValuesChange={[Function]}
        singleTap={[Function]}
      />
      <FormLabel>
        הערות
      </FormLabel>
      <FormInput
        editable={true}
        maxLength={40}
        multiline={true}
        normalizeFontSize={true}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="כתוב הערות נוספות כאן"
        value=""
      />
      <View
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <Button
          label="שלח"
          onPress={[Function]}
        />
        <Button
          label="חזור"
          onPress={[Function]}
        />
      </View>
    </ScrollViewMock>
  </View>
</Component>
`;
